### DISCUSSION TITLE
Prisma support for transaction rollback and Promise.all

### DISCUSSION THREAD

**Discussion by rubenlopezlozoya** [prisma/prisma#13564] [Q&A]
Hey team, we stumbled across an issue when trying to run Promise.all() within a $transaction and rolling back once a promise fails. This is what our implementation looks like:

```try{
prisma.$transaction((prisma) => async {
   const updates = items.map((item) => {
      prisma.updateMany(....)
   });
   await Promise.all(updates);
})
} catch (err) {
  // Handle the rollback...
}
```

if any of the promises rejects, the transaction will rollback but only all the operations that happened before the failing one. The rest of the operations that are being executed as part of the Promise.all() will still reach the database, leaving us with an undesired partial update. I would expect ORMs to handle this situation for me and somehow disregard/rollback all ongoing operations after one of the promises inside a transaction fails. I am wondering if what I want to achieve can be done using Prisma out of the box or if I will have to play around with Promise.allSettled API (or just going full sequential) to get what I want.

**Comment by aemartos** [comment]
We came across the same issue, is there an update here? TY!

**Comment by pantharshit00** [comment]
Hello @rubenlopezlozoya, our apologies for the delay in getting back to you! Due to the large number of questions in our community channels, unfortunately this slipped through the cracks. It's very important for us to address your question, so let us know in a comment if this is still an open discussion. 

I would suggest you to use the `$transaction([])` API here instead. There you can pass the array and it will run in a transaction without any partial updates to data.

**Reply by apalumbo** [reply]
@pantharshit00 Thank you for answering , I have another question related on this 

```
const [posts, totalPosts] = await prisma.$transaction([
  prisma.post.findMany({ where: { title: { contains: 'prisma' } } }),
  prisma.post.count(),
])
```
does the promises needs to be created just in the array?
Usually promises starts when created afaik , so I am wondering if , given the promises returned are PrismaPromise, this is handled somehow 

```
const posts = prisma.post.findMany({ where: { title: { contains: 'prisma' } } });
const count =   prisma.post.count();

const [posts, totalPosts] = await prisma.$transaction([
  posts,
  count,
])
```

**Reply by leppaott** [reply]
Prisma promises execute only when awaited or then()/catch() called etc, so yeah you can just map an array of prisma calls and then run them in the transaction.

**Reply by apalumbo** [reply]
@leppaott Thank you, it would be a good idea to add to the docs as AFAIK this is not a standard behaviour ;-) but a good one. I will try to submit a PR to the doc if possible

**Reply by pantharshit00** [reply]
Thanks for the follow up @leppaott

@apalumbo You can report the documentation issue here: https://github.com/prisma/docs/issues/new?assignees=&labels=docs&template=doc_request.md

**Reply by pantharshit00** [reply]
Hello again,

I'm following up to let you know that we added this information in our documentation:
<img width="692" alt="image" src="https://user-images.githubusercontent.com/22195362/204938170-68f50d72-22ef-47c6-bb82-e3a468fdf398.png">

Link: https://www.prisma.io/docs/concepts/components/prisma-client#3-use-prisma-client-to-send-queries-to-your-database

Thank you again for the feedback! In the future, don't hesitate to create a [Docs Issue](https://github.com/prisma/docs/issues/new?assignees=&labels=docs&template=doc_request.md) yourself - feedback and suggestions are always welcome!

**Reply by rj-au** [reply]
The latest docs are missing this important information now: https://www.prisma.io/docs/orm/prisma-client/queries/transactions#transaction-api

I had to find this discussion/thread to confirm I could create the promises before the $transaction.
~~I'm assuming the prisma queries have a **tick** timeout to allow the transaction code to be hit first, otherwise the query promises would have fired over the network already.~~ (See link below (about PrismaPromise(`thenable`)), for how it works)

**Reply by janpio** [reply]
This might be relevant @richard-stafflink: https://www.prisma.io/docs/orm/reference/prisma-client-reference#prismapromise-behavior

**Reply by rj-au** [reply]
Thanks - that makes it very clear - I didn't realise what I was looking for had a name.

**Comment by Azoom-Le-Van-Thuong** [comment]
@pantharshit00 pantharshit00
Hy Team, I have question about performance when using `$transaction`, 
The way 1 
`
await prisma.$transaction(async tx => { 
   await updateSomeThing()
   await createSomeThing()
})
`

The way 2 

`
await prisma.$transaction(async tx => { 
   await Promise.all([updateSomeThing,createSomeThing])
})
`

Which the best perfomance using transaction about time ?. thask. Instead of await, await I have Promise all. ? Pls help me

**Reply by apalumbo** [reply]
@Azoom-Le-Van-Thuong it should not related to prisma, but it should related to promises.

Them first one the `await createSomeThing()` promise is executed AFTER  updateSomeThing() is resolved , sequentially

The second the execution is in parallel  

you have to use the first way just when `createSomeThing` has some dependency (in data or db state) from `updateSomeThing`

**Comment by phineasla** [comment]
So just to be clear, we shouldn't use `await Promise.all()` in `$transaction()`? I can't use `$transaction([])` because I need to know which line failed and do some stuffs with it (non-Prisma related).
```
const users: User = [];
let failed: User;
await this.prismaService.$transaction(async (tx) => {
  await Promise.all(
    users.map((user) =>
      tx.users.create(...).catch((e) => {
        failed = user;
        throw e;
      })
    )
  );
});
```

**Reply by leppaott** [reply]
Not an expert, but seems this `BUG` hasn't been addressed yet. It can be worked around in some cases but of course you'd expect all calls through `tx` to be rollback despite calling with Promise.all.... Sure you can change that into a `for-loop` but is it a good fix?

**Reply by AlwanN01** [reply]
Use `Promise.allSettled` instead of `Promise.all`. `Promise.all` will immediately throw an error if any promise is rejected, without waiting for responses from other promises.
like this:
begin
update 2 OK
update 1 OK
update 3 Reject -> throw
rollback
update 5 OK <- BUG
update 4 OK

**Comment by sailonely** [comment]
I think the bug is caused by the behavior of Promise.all(), which rejects with the reason of the first rejected promise but does not stop the execution of other promises.
A similar case is described here: https://medium.com/@alkor_shikyaro/transactions-and-promises-in-node-js-ca5a3aeb6b74

**Comment by JonathanGafar** [comment]
Just curious...in the original post, shouldn't the Promise.all actually run sequentially, rather than in parallel, since each promise is a Prisma promise and there is only one connection available to use (since the Promise.all is within a transaction)?

**Reply by leppaott** [reply]
Prisma can use multiple connections see `connection_limit`.

**Reply by vincaslt** [reply]
The docs explicitly state that transactions will use a single connection, so presumably Promise.all has no effect? https://www.prisma.io/docs/orm/prisma-client/queries/transactions#using-transaction-within-promiseall

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
